runtime/wasm/src/host.rs

```diff
@@ -684,103 +689,138 @@ impl RuntimeHostTrait for RuntimeHost {
         log: Arc<Log>,
         state: BlockState,
     ) -> Box<Future<Item = BlockState, Error = Error> + Send> {
-        // Identify event handler for this log
-        let event_handler = match self.handler_for_log(&log) {
-            Ok(handler) => handler,
-            Err(e) => return Box::new(future::err(e)),
-        };
+        let logger = logger.clone();
+        let mapping_request_sender = self.mapping_request_sender.clone();
 
-        // Identify the event ABI in the contract
-        let event_abi = match util::ethereum::contract_event_with_signature(
-            &self.data_source_contract_abi.contract,
-            event_handler.event.as_str(),
-        ) {
-            Some(event_abi) => event_abi,
-            None => {
-                return Box::new(future::err(format_err!(
-                    "Event with the signature \"{}\" not found in \
-                     contract \"{}\" of data source \"{}\"",
-                    event_handler.event,
-                    self.data_source_contract_abi.name,
-                    self.data_source_name
-                )));
-            }
-        };
+        let block = block.clone();
+        let transaction = transaction.clone();
+        let log = log.clone();
 
-        // Parse the log into an event
-        let params = match event_abi.parse_log(RawLog {
-            topics: log.topics.clone(),
-            data: log.data.clone().0,
-        }) {
-            Ok(log) => log.params,
-            Err(e) => {
-                return Box::new(future::err(format_err!(
-                    "Failed to parse parameters of event: {}: {}",
-                    event_handler.event,
-                    e
-                )));
-            }
-        };
+        let data_source_name = self.data_source_name.clone();
+        let abi_name = self.data_source_contract_abi.name.clone();
+        let contract = self.data_source_contract_abi.contract.clone();
 
-        debug!(
-            logger, "Start processing Ethereum event";
-            "signature" => &event_handler.event,
-            "handler" => &event_handler.handler,
-            "data_source" => &self.data_source_name,
-            "address" => format!("{}", &log.address),
-        );
-
-        // Call the event handler and asynchronously wait for the result
-        let (result_sender, result_receiver) = oneshot::channel();
+        Box::new(
+            future::result(self.handlers_for_log(&log)).and_then(|event_handlers| {
+                stream::iter_ok(event_handlers).fold(state, move |state, event_handler| {
```

Comment:
I think we lost some readability in this function, this is using futures and streams event though it's synchronous, it doesn't express that only one event handler will actually be called, and it puts the entire function inside this loop.

A suggestion is to first somehow build a `Vec<(MappingEventHandler, Event)>`, gathering all the ABIs, and then get the correct `event_handler` (for example using [find_map](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find_map)), so the rest of the function can remain unchanged.

Reply:
I've updated the code, can you take another look?